deps/AIP-0002-kubernetes-deployment.md

```diff
@@ -0,0 +1,565 @@
+2. **Public Datasets**: DatasetManager downloads directly from URL (no upload needed)
+3. **Synthetic Datasets**: DatasetManager generates synthetic conversations
+
+The implementation will determine the optimal transfer method based on dataset size and type.
+
+### Dataset Distribution to Workers
+
+**Current Challenge:**
+- DatasetManager distributes conversations to worker pods via ZMQ DEALER/ROUTER
+- High request volume could create bottleneck at DatasetManager
+- ZMQ request/reply latency scales with worker count
+
+**Optimization Strategies (To Be Evaluated):**
+
+1. **Worker-Side Caching/Pre-Fetching**: Workers pre-fetch and cache conversations, request new ones less frequently
+2. **Batch Distribution**: DatasetManager sends batches of conversations per request instead of one-at-a-time
+3. **Multiple DatasetManager Replicas**: Scale DatasetManager horizontally with dataset sharding
+4. **Redis Cache**: DatasetManager caches conversations in Redis for workers to fetch
+5. **Shared Volumes**: DatasetManager mounts shared volumes to worker pods for dataset distribution
+
+**MVP Approach:**
+- Single DatasetManager pod with in-memory dataset
+- Workers request conversations via ZMQ DEALER/ROUTER as needed
+- Performance testing will determine if optimization is needed for high worker counts
+
+## Configuration Management
+
+### CLI Parameters
+**Kubernetes Deployment Parameters**:
+- `--kubernetes`: Enable Kubernetes deployment mode
+- `--kubernetes-namespace`: Target Kubernetes namespace (optional, defaults to auto-generated namespace)
+- `--kubeconfig`: Path to Kubernetes configuration file (optional, defaults to `~/.kube/config`)
+
+**Namespace Behavior**:
+- If `--kubernetes-namespace` is specified, AIPerf will use that namespace
+- If not specified, AIPerf will auto-generate a unique namespace (e.g., `aiperf-<timestamp>` or `aiperf-<job-id>`)
+- Auto-generated namespaces are automatically cleaned up after benchmark completion
+
+**Existing Parameters (Work with Both Deployment Modes)**:
+- `--record-processors`: Controls record processor scaling
+- `--workers-max` / `--max-workers`: Controls worker scaling
+
+## Container Images
+
+A single container image supports all AIPerf service modes. Service type is determined by environment variables passed to the pod, using the existing code to instantiate the appropriate service class.
+
+## Artifact and Export File Retrieval
+
+AIPerf generates output files including metrics exports (JSON, CSV) and logs that users need to access after benchmark completion. In the Kubernetes deployment, these files are generated by the Records Manager pod and must be retrieved to the user's local filesystem via the Kubernetes Python API.
+
+### Basic Approach
+
+The Records Manager pod acts as the central collection point for all benchmark artifacts. It receives processed metrics from distributed pods and generates the final export files in its local artifact directory using the existing AIPerf exporters.
```

Comment:
Are there incremental updates of the data so if a worker goes down mid run all of the data is not lost? If so are you writing this out to a persistent volume storage or streaming the data back to another location?